velox/functions/prestosql/window/FirstLastValue.cpp

```diff
@@ -82,9 +83,63 @@ class FirstLastValueFunction : public exec::WindowFunction {
     invalidRows_.applyToSelected([&](auto i) { rowNumbers_[i] = kNullRow; });
   }
 
+  void setRowNumbersRespectNulls(
+      const SelectivityVector& validRows,
+      const BufferPtr& frameStarts,
+      const BufferPtr& frameEnds) {
+    if constexpr (TValue == ValueType::kFirst) {
+      auto rawFrameStarts = frameStarts->as<vector_size_t>();
+      validRows.applyToSelected(
+          [&](auto i) { rowNumbers_[i] = rawFrameStarts[i]; });
+    } else {
+      auto rawFrameEnds = frameEnds->as<vector_size_t>();
+      validRows.applyToSelected(
+          [&](auto i) { rowNumbers_[i] = rawFrameEnds[i]; });
+    }
+  }
+
+  void setRowNumbersIgnoreNulls(
+      const SelectivityVector& validRows,
+      const BufferPtr& frameStarts,
+      const BufferPtr& frameEnds) {
+    auto extractNullsResult = partition_->extractNulls(
```

Comment:
nit: auto [firstFrame, frameSize] = partition_->extractNulls... for readability

Reply:
That shorthand didn't work for me. The lambda in applyToSelected(...) loop below couldn't resolve firstFrame if I wrote the code that way.